providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,764 @@
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        OPENSSL_free(key->propq);
+        key->propq = NULL;
+        if (!OSSL_PARAM_get_utf8_string(p, &key->propq, 0))
+            return 0;
+    }
+
+    if (publen != key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    return load_keys(key, pubenc, publen, NULL, 0);
+}
+
+static int mlx_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        if ((gctx->propq = OPENSSL_strdup(p->data)) == NULL)
+            return 0;
+    }
+    return 1;
+}
+
+static void *mlx_kem_gen_init(int v, OSSL_LIB_CTX *libctx, int selection,
+                              const OSSL_PARAM params[])
+{
+    static const int minimal =
+        OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS
+        | OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    PROV_ML_KEM_GEN_CTX *gctx = NULL;
+
+    /*
+     * We can only generate private keys, check that the selection is
+     * appropriate.
+     */
+    if (!ossl_prov_is_running()
+        || (selection & minimal) == 0
+        || (gctx = OPENSSL_zalloc(sizeof(*gctx))) == NULL)
+        return NULL;
+
+    gctx->variant = v;
+    gctx->libctx = libctx;
+    gctx->selection = selection;
+    if (mlx_kem_gen_set_params(gctx, params))
+        return gctx;
+
+    mlx_kem_gen_cleanup(gctx);
+    return NULL;
+}
+
+static const OSSL_PARAM *mlx_kem_gen_settable_params(ossl_unused void *vgctx,
+                                                     ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return settable;
+}
+
+static void *mlx_kem_gen(void *vgctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    MLX_KEY *key;
+    char *propq = gctx->propq;
+
+    if (gctx == NULL
+        || (gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) ==
+            OSSL_KEYMGMT_SELECT_PUBLIC_KEY)
+        return NULL;
+
+    /* Lose ownership of propq */
+    gctx->propq = NULL;
+    if ((key = mlx_kem_key_new(gctx->variant, gctx->libctx, propq)) == NULL)
+        return NULL;
+
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
+
+    /* XXX: Using same "propq" for all components */
```

Comment:
Why leave a comment like this here? If it is an issue to discuss then raise an issue.
I don't see an issue - this is being modelled as one algorithm - so one propq makes sense to me.
Absent and identified reason for two (and then how would two propqs make it down to the caller of this function then becomes the question).

Reply:
Playing devil's advocate: What if someone wanted to use different providers via different `propq`s for the different constituent algorithms? In such case, there should be a new (feature request) issue "Allow creation of hybrid algorithms (keys) using different `propqs`". But I'd leave that to people who want that feature.

For now, I'd argue this should be precluded and the impossibility of that should be documented rather prominently. Which actually raises the question: Where (in which .pod files) are these algorithms (even if only by reference to the draft standards they implement) actually documented?

Reply:
We'll address this issue when our ML-KEM is actually enabled in the FIPS provider, and we know what the FIPS status of the hybrids will be when the (2nd!) component is non-FIPS. This is supposedly allowed, and we may then need to figure out whether the second algorithm gets a propq, perhaps we should use a NULL prop for the ECX components (but the EC components are already FIPS-validated).  I am leaving the breadcrumb there. Though of course a GitHub issue is likely warranted.